feat(colors): add options for forcing color/no-color

thebanjomatic opened the pull request:
This change checks for --color / --no-color as command-line arguments as a mechanism for forcing color in non-TTY situations, or forcing --no-color for TTY. Additionally, this change hooks into the `FORCED_COLOR` environment variable that is used by the [`supports-color`](https://www.npmjs.com/package/supports-color) npm package which is used by a variety of packages in the node ecosystem.

The intention of this PR is to alleviate some of the concerns from Issue #897 by allowing users of turborepo to enable color output on CI runners such as Github Actions which support color output (as mentioned in the [fatih/color documentation](https://github.com/fatih/color/tree/master#github-actions)). When github actions and local builds are both running in color, then the cached output should be consistently in color, compared to the current situation where its a mix of color and non-color output with a remote cache shared between CI and users building interactively.

The `FORCED_COLOR` environment variable is something that I find myself already setting when running turbo so that the streamed output is in color for the build and testing tools that I am running with it, so it seemed like a natural way to provide this argument without needing to modify the CLI parameters being passed to turbo.

In the event multiple methods are being used, the precidence is:
* isTTY or not (provided by fatih/color)
* $NO_COLOR (if set to anything, then color is disabled; also provided by fatih/color)
* $FORCED_COLOR ("1", "2", "3", "true" will enable color, "0", or "false" will disable color; other values are ignored)
* --color/--no-color (the last positional argument wins if both exist)

thebanjomatic commented on cli/internal/ui/ui.go:
```diff
@@ -63,13 +57,34 @@ func Rainbow(text string) string {
+	
+	var outWriter, errWriter io.Writer
+
+	if color.NoColor {
+		outWriter = &stripAnsiWriter{file: os.Stdout}
+		errWriter = &stripAnsiWriter{file: os.Stderr}
+	} else {
+		outWriter = os.Stdout
+		errWriter = os.Stderr
+	}
```
We are only checking the value of color.NoColor on initialization so changing the value mid run isn't supported. This can be done more dynamically by moving the condition back inside of stripAnsiWriter and always returning that:
```go
func (into *stripAnsiWriter) Write(src []byte) (n int, err error) {
  if color.NoColor {
    src = ansiRegex.ReplaceAll(src, []byte{})
  }
	into.file.Write(src)
	return n, nil
}
```

But this way avoids adding the extra condition every time text gets output. I don't have a strong preference here and I haven't profiled to know what the impact of that would be.

thebanjomatic commented on cli/internal/ui/ui.go:
```diff
@@ -50,7 +44,7 @@ func rgb(i int) (int, int, int) {
 // Rainbow function returns a formated colorized string ready to print it to the shell/terminal
 func Rainbow(text string) string {
-	if !IsTTY {
+	if color.NoColor {
```
I think I would prefer if the `applyForcedColor` function returned the value of `color.NoColor` as a `bool` instead of leaking this implementation detail as an implementation detail. This approach would only make sense if we stick with the current implementation where the value is only checked once on construction of `ColoredUi`.

thebanjomatic replied:
While we don't technically need to check this, the code is doing a non-trivial amount of formatting looping over each character passed in so I can go either way, but I like not having to check color.NoColor and by removing it we can prevent that abstraction leak as well.

thebanjomatic commented on .github/workflows/ci-go.yml:
```diff
@@ -41,10 +41,10 @@ jobs:
 
       - name: Build & Unit Test
-        run: pnpm turbo run test --scope=cli
+        run: pnpm turbo -- run test --scope=cli --color
```
Only the positional arguments were being passed in before:
```
./turbow.sh "run" "test"
```
This now gets sent as:
```
./turbow.sh "run" "test" "--scope=cli" "--color"
```

Likely a change in behavior when switching to pnpm from yarn

thebanjomatic commented:
The unit tests should all be fixed now, and I have also enabled the color output for the ci workflow in github actions for this repo:
![image](https://user-images.githubusercontent.com/17907922/161863468-c15f69e5-3fa4-4236-8d95-6fbe7dfae3e5.png)

gsoltis commented on cli/internal/ui/colors.go:
```diff
@@ -0,0 +1,39 @@
+	"github.com/fatih/color"
+)
+
+type ForcedColor int
```
If we go with `ColorMode`, I think that would give us `GetColorModeFromEnv` and `applyColorMode` below, as well as `ColorModeUndefined`. The others I'm a little less certain: maybe `ColorModeForced`, `ColorModeSuppressed`?

thebanjomatic replied:
Went with the suggested ColorModeForced/ColorModeSuppressed

gsoltis reviewed (changes requested):
I like where this is going!

thebanjomatic commented on cli/internal/ui/colors.go:
```diff
@@ -0,0 +1,39 @@
+type ForcedColor int
+
+const (
+	Force_Color_Undefined ForcedColor = iota
```
I don't think we are actually using the int values for anything, so it should be fine to start at `iota + 1`.

thebanjomatic commented on cli/internal/ui/colors.go:
```diff
@@ -0,0 +1,39 @@
+)
+
+func GetForcedColorFromEnv() ForcedColor {
+	switch forcedColor := os.Getenv("FORCED_COLOR"); {
```
Good thinking, same with the function that applies the specific values of 0,1,2,3.

thebanjomatic commented on cli/internal/ui/ui.go:
```diff
@@ -63,13 +57,34 @@ func Rainbow(text string) string {
 }
 
+type stripAnsiWriter struct {
+	file *os.File
```
This can be an io.Writer now. A previous incarnation of this code was using WriteString as well but that wound up being uneccessary.

thebanjomatic commented:
Thanks for your review @gsoltis I believe I have addressed all of your feedback. I mostly used your suggested naming, I think the only thing that was unspecified was the function to construct the `cli.ColoredUi` instance. For that, I went with `Default()` and `BuildColoredUi(colorMode ColorMode)` where `Default()` just returns `BuildColorMode(ColorModeUndefined)`.

gsoltis commented on cli/cmd/turbo/main.go:
```diff
@@ -28,6 +28,7 @@ func main() {
 	traceFile := ""
 	cpuprofileFile := ""
 	argsEnd := 0
+	forceColor := uiPkg.GetColorModeFromEnv()
```
`forceColor` -> `colorMode`

gsoltis commented on cli/internal/ui/ui.go:
```diff
@@ -63,13 +54,38 @@ func Rainbow(text string) string {
+}
+
+func (into *stripAnsiWriter) Write(src []byte) (n int, err error) {
+	into.wrappedWriter.Write(ansiRegex.ReplaceAll(src, []byte{}))
```
This looks like it is losing the return values?

I think in general we prefer not to use named return variables for this reason and general readability.

gsoltis reviewed (changes requested):
Awesome, just a couple minor things and I think this should be good to go

thebanjomatic commented:
I also made an attempt at writing some [documentation](https://github.com/vercel/turborepo/pull/990/files#diff-a8cc69287291b3e84bd88860e91ce531c711d2735334e48a96235900ec98bbd2R28-R67) for the new flags

gsoltis reviewed (approved):
Awesome, thanks for submitting this!
